### PR TITLE
Bug 1803465 - extend isURLLenient to match IPv6 literals

### PR DESCRIPTION
**Pull request by pmarks-net**
This pull request lets Firefox for Android accept IPv6 literals typed into the address bar.

For example: <a href="https://[2606:4700:4700::1111]/">https://[2606:4700:4700::1111]/</a>

The bug was originally reported as https://github.com/mozilla-mobile/fenix/issues/4343, but @YoRyan's https://github.com/mozilla-mobile/android-components/pull/5546 was rejected because it grew [a big slow regex](https://github.com/mozilla-mobile/android-components/issues/5594).

The big regex [no longer exists](https://github.com/mozilla-mobile/android-components/commit/f9282686cd8243dc3faead1cdeedbf2a73dd1215) (yay!) leaving this 1-line `isURLLenient` regex as the final roadblock.

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1803465
https://bugzilla.mozilla.org/show_bug.cgi?id=1807752

### CODE REVIEW

**Comment by pmarks-net** [comment]
@csadilek and @mcomella originally wrote this regex.

**Comment by pmarks-net** [comment]
I tested this using https://firefox-source-docs.mozilla.org/mobile/android/fenix.html and `./gradlew support-utils:test`.

I'm not sure if the 3 workflow failures are my fault, nor how to rerun them.

**Comment by Djfe** [comment]
great, thank you for working on this 👍
Would it make sense to also introduce some assertFalse?
like for misspelled ipv6 style urls?

**Comment by pmarks-net** [comment]
@Djfe this regex has no concept of a "misspelled" IP address.  It's not looking at the address itself, just very broad character classes to distinguish possible URLs from garbage.

**Comment by pmarks-net** [comment]
Added more test cases.  When I type these into the UI, it either shows "Invalid Address" or a completely blank tab, depending on the mood of the downstream URL parser.

Note that both of these states are already reachable in Firefox 118:

- `http://x:x` -> "Invalid Address"
- `x://x` -> blank tab

**Comment by pmarks-net** [comment]
I'm still waiting for a reviewer.  @csadilek perhaps?

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:123]: I think this block includes tests that assert that `isURLLike` works as expected (returning false for malformed IPv6 literals), as well as tests showing the limits of the current regex (returning the wrong result for wellformed and malformed IPv6 literals). 

It would be a bit clearer to separate the two types of tests in two groups described with a comment.

Comment (titooan) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:123]: Also, we noticed that we could add tests with an URL with the wrong number of bytes : 
```kotlin 
        assertFalse(isURLLike("https://[1:2:3:4:5:6:7]/"))
        assertFalse(isURLLike("https://[1:2:3:4:5:6:7:8:9]/"))
```

Those test are currently failing. Would it be possible to adapt the regex so that we take this case into account? 

If not, we can put those tests in the block of tests showing the limits of the current regex.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:103]: Here is a suggestion that I think makes it easier to test several variations of each "base IPv6 literal", including `http:`, `https:`, trailing slash... 
```suggestion
        val validBaseUrls = listOf(
            "[::]",
            "[::1]",
            "[1::]",
            "[::]/",
            "[1:2:3:4:5:6:7:8]",
            "[2001:db8::1.2.3.4]",
        )

        validBaseUrls.forEach { url ->
            assertTrue(isURLLike(url))
            assertTrue(isURLLike("$url/"))
            assertTrue(isURLLike("https://$url"))
            assertTrue(isURLLike("https://$url/"))
            assertTrue(isURLLike("http://$url"))
            assertTrue(isURLLike("http://$url/"))
        }
```

What do you think?

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:66]: Can we also add some more similar test coverage from UrlStringUtils into WebUrlFinderTest?

**Review by titooan — CHANGES_REQUESTED** [review]
Hi pmarks-net,
Thanks a lot for this PR and sorry for the late review. 

While we were reviewing this, we tried to think of all the cases and here are some suggestions we came up with, that you can find in this [patch file](https://github.com/mozilla-mobile/firefox-android/files/13378157/PR-4090.patch).


Please feel free to ask for clarifications is anything is not clear enough.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: This regex makes no attempt to parse IP addresses, so cases like `[1:2:3:4:5:6:7]` belong in the "Invalid IPv6 literals incorrectly recognized as valid" bucket.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: Done.  I removed "[::]/" because the / is tested later, added "[::1]:8080" for coverage, and added tests for "https:[::]"-style URLs.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:66]: Added `isUrlLikeEmulated()`, copied all the test cases, and marked cases that differ with `INVERT()`.

**Review by pmarks-net — COMMENTED** [review]
These changes were incorporated into  commit 674be27070f77e1d6e55b6bd67f0f456f994bb35

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (titooan) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: Good catch!

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:211]: I'm sorry, I think my comment about adding tests for WebUrlFinderTest was not clear. 

Here is what I had in mind: 
```kotlin 
    @Test
    fun testIPv6LiteralsInsideSentence() {
        val validIPv6Literals = listOf(
            "[::]",
            "[::1]",
            "[1::]",
            "[1:2:3:4:5:6:7:8]",
            "[2001:db8::1.2.3.4]",
            "[::1]:8080",
        ).flatMap { url ->
            listOf(
                "https://$url",
                "https://$url/",
                "https:$url",
                "http://$url",
                "http://$url/",
                "http:$url",
            )
        }

        validIPv6Literals.forEach { url ->
            assertEquals(url, find("random_text $url other_random_text"))
        }
    }
```

The idea would be to make sure the regex is able to find the IPv6 literals inside a sentance containing them. 
What do you think?

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:109]: I realize we could also add: 
```kotlin
            assertTrue(isURLLike("http:$url/"))
            assertTrue(isURLLike("https:$url/"))
```

**Review by titooan — APPROVED** [review]

**Review by titooan — CHANGES_REQUESTED** [review]
Thanks a lot for the corrections you made! 
I added a few suggestions, please feel free to ask if anything is unclear.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:129]: My `isUrlLikeEmulated()` tests indicate that "fe80::" and "x:[" belong in a different category, because the URI class thinks they are valid.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt]: I've updated isUrlLikeEmulated() to use `find("random_text $it other_random_text")`, so it covers the cases you listed and more.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt:110]: done

**Comment by pmarks-net** [comment]
GitHub says "Merging is blocked - Merging can be performed automatically once the requested changes are addressed", so I'm going to try the "Re-request review" button.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (titooan) [inline android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:139]: Sorry, I forgot to specify that in my previous comment: we are adding complexity with this annotation-like function, that I think we could replace with clear comments around the tests.

We could replace the use of `assertTrue(INVERT(isURLLike(xxx)))` by `assertFalse(isURLLike(xxx))`. Those asserts could be gathered with a comment above explaining that the ideal regex should have the same result as the one in `UrlStringUtils.kt`, but the current one gives a different result. Also explaining that we         allow it for now, until bug 1685152 is fixed.

What do you think?

**Review by titooan — CHANGES_REQUESTED** [review]
Thanks a lot for those fixes! 

Let's just fix one last comment, and it's good to ship.

File: android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt
Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:139]: > replace the use of assertTrue(INVERT(isURLLike(xxx))) by assertFalse(isURLLike(xxx)). Those asserts could be gathered with a comment

I intentionally kept assertTrue/assertFalse the same as the original tests, and replaced the comments with a programmatic annotation, in order to reduce the cognitive effort required to maintain a second copy.  Your suggestion would *increase* complexity when both functions are considered.

Comment (titooan) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:139]: I do understand the idea and the value of trying to keep both tests as similar as possible.
Nevertheless, this is starting a new pattern in the codebase, and maybe we should try to follow the existing conventions instead?

Maybe we could get a second opinion from someone else? cc: @jonalmeida

Comment (jonalmeida) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt:139]: Hey! I agree with @titooan, I too would prefer avoiding new patterns. Adding these cases that should pass/fail can be in a separate section with a well-worded comment to explain that these should be have differently when the linked bug is fixed.

> I intentionally kept assertTrue/assertFalse the same as the original tests, and replaced the comments with a programmatic annotation, in order to reduce the cognitive effort required to maintain a second copy. Your suggestion would increase complexity when both functions are considered.

The felt complexity of both solutions are indeed subjective. In cases like this, I would recommend the advice of the code maintainers to decide what is the more preferred way, especially in terms of conventions the codebase follows. 🙂

Comment (pmarks-net) [reply android-components/components/support/utils/src/test/java/mozilla/components/support/utils/WebURLFinderTest.kt]: done.

**Comment by titooan** [comment]
bors try

**Comment by jonalmeida** [comment]
> bors try

It looks like the CI we depend on for our contributor PRs is down, so I've pushed [a separate branch](https://github.com/mozilla-mobile/firefox-android/pull/4566) to run all our CI tests.

**Comment by jonalmeida** [comment]
> > bors try
> 
> It looks like the CI we depend on for our contributor PRs is down, so I've pushed [a separate branch](https://github.com/mozilla-mobile/firefox-android/pull/4566) to run all our CI tests.

CI has passed! I've pushed a small update to correct the changelog entry since we've cut the previous version already. Let's land this!

**Review by titooan — APPROVED** [review]
Thanks a lot for your contribution @pmarks-net, it looks pefect to me! 
Let's land it!

**Comment by pmarks-net** [comment]
Firefox 122 (with the IPv6 literals fix) is now live on the Play store.

I'm mentioning https://github.com/mozilla-mobile/fenix/issues/4343 to notify the original reporters.

**Pull request merged**
